src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonerCreationService.kt

```diff
@@ -95,28 +95,28 @@ class PrisonerCreationService(
     return PrisonerIdentifier().withId(currentSequence.prisonerIdentifier)
   }
 
-  private fun gender(code: String): Result<Gender> =
-    genderRepository.findByIdOrNull(ReferenceCode.Pk(Gender.SEX, code))?.let { success(it) } ?: failure(
-      EntityNotFoundException.withMessage("Gender $code not found")
-    )
+  private fun gender(code: String): Result<Gender> = genderRepository.getMandatoryReference(code, Gender.SEX)
 
-  private fun ethnicity(code: String?): Result<Ethnicity>? =
-    code?.takeIf { it.isNotBlank() }?.let {
-      ethnicityRepository.findByIdOrNull(ReferenceCode.Pk(Ethnicity.ETHNICITY, code))?.let { success(it) }
-        ?: failure(EntityNotFoundException.withMessage("Ethnicity $code not found"))
-    }
+  private fun title(code: String?): Result<Title?> = titleRepository.getOptionalReference(code, Title.TITLE)
 
-  private fun title(code: String?): Result<Title>? =
-    code?.takeIf { it.isNotBlank() }?.let {
-      titleRepository.findByIdOrNull(ReferenceCode.Pk(Title.TITLE, code))?.let { success(it) }
-        ?: failure(EntityNotFoundException.withMessage("Title $code not found"))
-    }
+  private fun suffix(code: String?): Result<Suffix?> = suffixRepository.getOptionalReference(code, Suffix.SUFFIX)
+
+  private fun ethnicity(code: String?): Result<Ethnicity?> = ethnicityRepository.getOptionalReference(code, Ethnicity.ETHNICITY)
+
+  private inline fun <reified T: ReferenceCode> ReferenceCodeRepository<T>.getMandatoryReference(code: String, pk: String): Result<T> =
+    findByIdOrNull(ReferenceCode.Pk(pk, code))
+      ?.let { success(it) }
+      ?: failure(EntityNotFoundException.withMessage("${T::class.java} $code not found"))
+
+  private inline fun <reified T: ReferenceCode> ReferenceCodeRepository<T>.getOptionalReference(code: String?, pk: String): Result<T?> =
```

Comment:
These extension functions could be pushed up to ReferenceCode to make them available everywhere. 

You could probably also do something clever with the pk and derive it from T. Maybe make ReferenceCode a sealed class and have a static function mapping type to PK. Or summat like that.